### PR TITLE
Limit the number of concurrent requests per node

### LINKED ISSUE #31192: Revise Default max concurrent search requests
Spinoff from #31171

We today use `5 * numNodes` in the cluster as the default number of concurrent search requests issues by a single high level request. This has several issues ie. in the CCS case. There is also valid arguments that this number is arbitrary. Lets discuss this and find a good default that also works for CCS

### PR DESCRIPTION
**Pull request by s1monw**
With `max_concurrent_shard_requests` we used to throttle / limit
the number of concurrent shard requests a high level search request
can execute per node. This had several problems since it limited the
number on a global level based on the number of nodes. This change
now throttles the number of concurrent requests per node while still
allowing concurrency across multiple nodes.

Closes #31192

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/action/search/InitialSearchPhase.java [hunk 1]
```diff
@@ -197,6 +192,53 @@ public boolean isForceExecution() {
...
+            this.semaphore = new Semaphore(permits);
+        }
+        void finishAndRunNext() {
+            semaphore.release();
+            tryRun();
+        }
+
+        synchronized void add(Runnable runnable) {
+            if (queue == null) { // create this lazily
+                queue = new LinkedList<>();
```
Comment (jpountz) [inline server/src/main/java/org/elasticsearch/action/search/InitialSearchPhase.java]: not that it matters much but ArrayDeque should be better?

File: server/src/main/java/org/elasticsearch/action/search/InitialSearchPhase.java [hunk 2]
```diff
@@ -197,6 +192,53 @@ public boolean isForceExecution() {
...
+        PendingExecutions(int permits) {
+            this.semaphore = new Semaphore(permits);
+        }
+        void finishAndRunNext() {
+            semaphore.release();
+            tryRun();
+        }
+
+        synchronized void add(Runnable runnable) {
+            if (queue == null) { // create this lazily
```
Comment (jpountz) [inline server/src/main/java/org/elasticsearch/action/search/InitialSearchPhase.java]: why does it need to be lazy?

File: server/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java [hunk 3]
```diff
@@ -258,7 +254,7 @@ protected void doExecute(Task task, SearchRequest searchRequest, ActionListener<
                 }
                 String[] aliases = aliasFilter.getAliases();
                 String[] finalIndices = aliases.length == 0 ? new String[] {shardId.getIndexName()} : aliases;
-                // here we have to map the filters to the UUID since from now on we use the uuid for the lookup
+                // here we have to pendingExecutionsPerNode the filters to the UUID since from now on we use the uuid for the lookup
```
Comment (jpountz) [inline server/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java]: search/replace issue?

File: server/src/main/java/org/elasticsearch/action/search/InitialSearchPhase.java [hunk 4]
```diff
@@ -209,25 +251,52 @@ private void performPhaseOnShard(final int shardIndex, final SearchShardIterator
...
-                    public void innerOnResponse(FirstResult result) {
-                        maybeFork(thread, () -> onShardResult(result, shardIt));
-                    }
+            final PendingExecutions pendingExecutions = throttleConcurrentRequests ? pendingExecutionsPerNode.computeIfAbsent(shard.currentNodeId(),
+                n -> new PendingExecutions(maxConcurrentRequestsPerNode)) : null;
+            Runnable r = () -> {
+                try {
+                    executePhaseOnShard(shardIt, shard, new SearchActionListener<FirstResult>(new SearchShardTarget(shard
+                        .currentNodeId(),
+                        shardIt.shardId(), shardIt.getClusterAlias(), shardIt.getOriginalIndices()), shardIndex) {
```
Comment (jpountz) [inline server/src/main/java/org/elasticsearch/action/search/InitialSearchPhase.java]: indentation of the two above lines doesn't help readability

File: server/src/main/java/org/elasticsearch/action/search/SearchResponse.java [hunk 5]
```diff
@@ -199,7 +199,7 @@ public void scrollId(String scrollId) {
      * If profiling was enabled, this returns an object containing the profile results from
      * each shard.  If profiling was not enabled, this will return null
      *
-     * @return The profile results or an empty map
+     * @return The profile results or an empty pendingExecutionsPerNode
```
Comment (jpountz) [inline server/src/main/java/org/elasticsearch/action/search/SearchResponse.java]: search/replace issue?

File: server/src/main/java/org/elasticsearch/action/search/SearchResponseSections.java [hunk 6]
```diff
@@ -90,7 +90,7 @@ public final int getNumReducePhases() {
 
     /**
      * Returns the profile results for this search response (including all shards).
-     * An empty map is returned if profiling was not enabled
+     * An empty pendingExecutionsPerNode is returned if profiling was not enabled
```
Comment (jpountz) [inline server/src/main/java/org/elasticsearch/action/search/SearchResponseSections.java]: search/replace issue?

File: server/src/main/java/org/elasticsearch/action/search/SearchTransportService.java [hunk 7]
```diff
@@ -203,8 +203,8 @@ public RemoteClusterService getRemoteClusterService() {
     }
 
     /**
-     * Return a map of nodeId to pending number of search requests.
-     * This is a snapshot of the current pending search and not a live map.
+     * Return a pendingExecutionsPerNode of nodeId to pending number of search requests.
```
Comment (jpountz) [inline server/src/main/java/org/elasticsearch/action/search/SearchTransportService.java]: search/replace issue?

File: server/src/main/java/org/elasticsearch/action/search/InitialSearchPhase.java [hunk 8]
```diff
@@ -209,25 +251,52 @@ private void performPhaseOnShard(final int shardIndex, final SearchShardIterator
...
-                });
-            } catch (final Exception e) {
-                /*
-                 * It is possible to run into connection exceptions here because we are getting the connection early and might run in to
-                 * nodes that are not connected. In this case, on shard failure will move us to the next shard copy.
-                 */
-                fork(() -> onShardFailure(shardIndex, shard, shard.currentNodeId(), shardIt, e));
+                }
+            };
+            if (pendingExecutions == null || pendingExecutions.tryAcquire()) {
```
Comment (jpountz) [inline server/src/main/java/org/elasticsearch/action/search/InitialSearchPhase.java]: maybe remove this tryAcquire and go through the else block directly? I think it's simpler if the management of the semaphore permits are contained in PendingExecutions?

**Review by jpountz — CHANGES_REQUESTED** [review]
It looks like you had a search/replace issue but I like it in general. Given that it is an expert option I don't think we need sophisticated backward compatibility, let's just add a note about it in the breaking changes for 7.0?

**Comment by s1monw** [comment]
@jpountz I addressed your comments

**Review by jpountz — APPROVED** [review]
LGTM

**Pull request merged**